src/main/java/io/vertx/core/impl/utils/CyclicSequence.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright (c) 2011-2017 Contributors to the Eclipse Foundation
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+ * which is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+ */
+
+package io.vertx.core.impl.utils;
+
+import java.util.*;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * A thread safe cyclic sequence of elements that can be used for round robin.
+ * <p/>
+ * A sequence is immutable and mutations uses {@link #add(Object)} and {@link #remove(Object)}
+ * to return a modified copy of the current instance.
+ * <p/>
+ * The iterator uses a volatile index, so it can be incremented concurrently by several
+ * threads with locking.
+ *
+ * @author <a href="http://tfox.org">Tim Fox</a>
+ * @author <a href="mailto:julien@julienviet.com">Julien Viet</a>
+ */
+public class CyclicSequence<T> implements Iterable<T>, Iterator<T> {
```

Comment:
This is a critical piece in EventBus implementation. Why changing completely, especially if you want to backport it in 3.5.2 ?

Reply:
I can make it internal again but then I have to remove the corresponding CyclicSequence test because it requires to use HandlerHolder to test it.

Reply:
It is changed: before it only implemented `Iterable`, not it also implements `Iterator`.